File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/config/WebServerConfig.java
package com.github.arhor.simple.expense.tracker.config;

import lombok.RequiredArgsConstructor;
import lombok.val;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.ResourceLoader;
import org.springframework.core.io.support.ResourcePatternResolver;
import org.springframework.core.io.support.ResourcePatternUtils;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.config.annotation.PathMatchConfigurer;
import org.springframework.web.servlet.config.annotation.ResourceHandlerRegistry;
import org.springframework.web.servlet.config.annotation.ViewControllerRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

import com.github.arhor.simple.expense.tracker.config.properties.ApplicationProps;

import static org.springframework.web.method.HandlerTypePredicate.forAnnotation;

@Configuration(proxyBeanMethods = false)
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class WebServerConfig implements WebMvcConfigurer {

    private final ApplicationProps applicationProps;

    @Override
    public void addViewControllers(final ViewControllerRegistry registry) {
        registry.addViewController("/{path:[^\\.]*}").setViewName("forward:/");
    }

    @Override
    public void configurePathMatch(final PathMatchConfigurer configurer) {
        applicationProps.apiPathPrefix().ifPresent(pathPrefix -> {
            configurer.addPathPrefix(pathPrefix, forAnnotation(RestController.class));
        });
    }

    @Override
    public void addResourceHandlers(final ResourceHandlerRegistry registry) {
        applicationProps.resources().ifPresent(it -> {
            val patterns = it.patterns().toArray(String[]::new);
            val location = it.location().toArray(String[]::new);

            registry.addResourceHandler(patterns).addResourceLocations(location);
        });
    }

    @Bean
    @ConditionalOnMissingBean
    public ResourcePatternResolver resourcePatternResolver(final ResourceLoader resourceLoader) {
        return ResourcePatternUtils.getResourcePatternResolver(resourceLoader);
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/config/WebServerConfig.kt
package com.github.arhor.simple.expense.tracker.config

import com.github.arhor.simple.expense.tracker.config.props.ApplicationProps
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.ResourceLoader
import org.springframework.core.io.support.ResourcePatternResolver
import org.springframework.core.io.support.ResourcePatternUtils
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.method.HandlerTypePredicate
import org.springframework.web.servlet.config.annotation.PathMatchConfigurer
import org.springframework.web.servlet.config.annotation.ResourceHandlerRegistry
import org.springframework.web.servlet.config.annotation.ViewControllerRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

@Configuration(proxyBeanMethods = false)
class WebServerConfig(private val applicationProps: ApplicationProps) : WebMvcConfigurer {

    override fun addViewControllers(registry: ViewControllerRegistry) {
        registry.addViewController("/{path:[^\\.]*}").setViewName("forward:/")
    }

    override fun configurePathMatch(configurer: PathMatchConfigurer) {
        applicationProps.apiPathPrefix?.let {
            configurer.addPathPrefix(it, HandlerTypePredicate.forAnnotation(RestController::class.java))
        }
    }

    override fun addResourceHandlers(registry: ResourceHandlerRegistry) {
        applicationProps.resources?.let {
            val patterns = it.patterns.toTypedArray()
            val location = it.location.toTypedArray()

            registry.addResourceHandler(*patterns).addResourceLocations(*location)
        }
    }

    @Bean
    @ConditionalOnMissingBean
    fun resourcePatternResolver(resourceLoader: ResourceLoader): ResourcePatternResolver {
        return ResourcePatternUtils.getResourcePatternResolver(resourceLoader)
    }
}
